python/ray/data/_internal/compute.py

```diff
@@ -425,21 +463,47 @@ def _map_block_split(
 
 
 def _map_block_nosplit(
-    block: Block,
     block_fn: BlockTransform,
     input_files: List[str],
     fn: Optional[UDF],
-    *fn_args,
+    num_blocks: int,
+    *blocks_and_fn_args: Union[Block, Any],
     **fn_kwargs,
 ) -> Tuple[Block, BlockMetadata]:
     stats = BlockExecStats.builder()
     builder = DelegatingBlockBuilder()
+    blocks, fn_args = blocks_and_fn_args[:num_blocks], blocks_and_fn_args[num_blocks:]
     if fn is not None:
         fn_args = (fn,) + fn_args
-    for new_block in block_fn(block, *fn_args, **fn_kwargs):
+    for new_block in block_fn(blocks, *fn_args, **fn_kwargs):
         builder.add_block(new_block)
     new_block = builder.build()
     accessor = BlockAccessor.for_block(new_block)
     return new_block, accessor.get_metadata(
         input_files=input_files, exec_stats=stats.build()
     )
+
+
+def _bundle_blocks_up_to_size(
+    blocks: List[Tuple[Block, BlockMetadata]],
+    target_size: int,
+) -> List[Tuple[List[Block], List[BlockMetadata]]]:
+    """Group blocks into bundles that are up to (but not exceeding) the provided target
+    size.
+    """
+    block_bundles = []
+    curr_bundle = []
+    curr_bundle_size = 0
+    for b, m in blocks:
+        num_rows = m.num_rows
+        if num_rows is None:
+            num_rows = float("inf")
+        if curr_bundle_size > 0 and curr_bundle_size + num_rows > target_size:
+            block_bundles.append(curr_bundle)
+            curr_bundle = []
+            curr_bundle_size = 0
+        curr_bundle.append((b, m))
+        curr_bundle_size += num_rows
+    if curr_bundle:
+        block_bundles.append(curr_bundle)
+    return [tuple(zip(*block_bundle)) for block_bundle in block_bundles]
```

Comment:
shall we add a warning if any of `block_bundle` has too much data in terms of bytes size (say more than 2X of `context.target_max_block_size`)?

we may also want to add a warning if number of `block_bundle` is much smaller than number of original blocks (say 10X reduction). This means we reduce number of tasks (parallelism) by too much, and this might bring some surprising slowness to users if they run heavy computation per row/batch.

Reply:
>shall we add a warning if any of block_bundle has too much data in terms of bytes size (say more than 2X of context.target_max_block_size)?

Good call out! I think that the `batch_size` being larger than `context.target_max_block_size` would not be common, but it definitely could happen, especially with the default batch size of 4096 for very wide tables or very large images. I think logging a warning in that case would be a good idea, I'll add that!

>we may also want to add a warning if number of block_bundle is much smaller than number of original blocks (say 10X reduction). This means we reduce number of tasks (parallelism) by too much, and this might bring some surprising slowness to users if they run heavy computation per row/batch.

Hmm we'd need to have a good recommendation for what the user could do to remedy this... Maybe something like:
```python
if len(block_bundle) > 10:
    logger.warning(
        f"Having to send {len(block_bundle)} blocks to a single ds.map_batches() task to create a "
        f"batch of size {batch_size}, which may result in less transformation parallelism than "
        "expected. This may indicate that your blocks are too small and/or your batch size is too "
        "large, and you may want to decrease your read parallelism and/or decrease your batch size."
    )
```
How does that sound?

Reply:
The error message sounds good, for the `if` condition, I am more towards to check overall parallelism:

```
if len(blocks) / len(block_bundles) > 10:
  ...
```